src/erc721/ERC721.sol

```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity >=0.8.0;
+
+/// @notice Modern and gas efficient ERC-721 + EIP-2612-like implementation.
+// ! TO DO - add safeTransfer stuff 🙏
+contract ERC721 {
+    /*///////////////////////////////////////////////////////////////
+                                  EVENTS
+    //////////////////////////////////////////////////////////////*/
+    
+    event Transfer(address indexed from, address indexed to, uint256 indexed tokenId);
+    
+    event Approval(address indexed owner, address indexed spender, uint256 indexed tokenId);
+    
+    event ApprovalForAll(address indexed owner, address indexed operator, bool approved);
+    
+    /*///////////////////////////////////////////////////////////////
+                             METADATA STORAGE
+    //////////////////////////////////////////////////////////////*/
+    
+    string public name;
+    
+    string public symbol;
+    
+    /*///////////////////////////////////////////////////////////////
+                             ERC-721 STORAGE
+    //////////////////////////////////////////////////////////////*/
+    
+    uint256 public totalSupply;
+    
+    mapping(address => uint256) public balanceOf;
+    
+    mapping(uint256 => address) public ownerOf;
+    
+    mapping(uint256 => string) public tokenURI;
+    
+    mapping(uint256 => address) public getApproved;
+ 
+    mapping(address => mapping(address => bool)) public isApprovedForAll;
+    
+    /*///////////////////////////////////////////////////////////////
+                         PERMIT/EIP-2612 STORAGE
+    //////////////////////////////////////////////////////////////*/
+    
+    bytes32 public constant PERMIT_TYPEHASH =
+        keccak256("Permit(address owner,address spender,uint256 tokenId,uint256 nonce,uint256 deadline)");
+        
+    bytes32 public immutable DOMAIN_SEPARATOR;
+
+    mapping(address => uint256) public nonces;
+    
+    constructor(
+        string memory _name,
+        string memory _symbol
+    ) {
+        name = _name;
+        symbol = _symbol;
+
+        DOMAIN_SEPARATOR = keccak256(
+            abi.encode(
+                keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"),
+                keccak256(bytes(name)),
+                keccak256(bytes("1")),
+                block.chainid,
+                address(this)
+            )
+        );
+    }
+    
+    /*///////////////////////////////////////////////////////////////
+                              ERC-721 LOGIC
+    //////////////////////////////////////////////////////////////*/
+    
+    function supportsInterface(bytes4 sig) external pure returns (bool) {
+        return (sig == 0x80ac58cd || sig == 0x5b5e139f);
```

Comment:
is this required?

Reply:
yeah, though I find it tedious hah (like anyone can claim their contract supports an interface without actually implementing....) But to the point ~~ [ERC-721](https://eips.ethereum.org/EIPS/eip-721) supports [ERC-165](https://eips.ethereum.org/EIPS/eip-165) 'standard interface detection', 
![image](https://user-images.githubusercontent.com/41117279/132996318-9023a104-c4fc-485f-b86d-77a8fb198504.png)

Reply:
here, we flag that ERC721 implementation supports `ERC721` and "metadata extension" 
![image](https://user-images.githubusercontent.com/41117279/132996394-4f853388-636d-48f0-ba7c-a88fafa9bf9b.png)

ERC721 - `0x80ac58cd`
![image](https://user-images.githubusercontent.com/41117279/132996462-7926ec9b-a6cc-48ae-bf20-181b635f27dd.png)

Metadata - `0x5b5e139f`
![image](https://user-images.githubusercontent.com/41117279/132996482-efeeec63-ab08-4650-8e92-1e724f99a8c0.png)

This implementation also partially supports "enumeration extension" (by ref to `totalSupply()`, but I have not flagged that in ERC-165 interface, as the indexing functions don't seem all that useful, and have not included,
![image](https://user-images.githubusercontent.com/41117279/132996413-c6133877-837f-4e7c-8c19-6bc1dc5fce01.png)

Reply:
I think you are also supposed to add the ERC165 interface too actually, to be totally compliant with 165

```solidity
contract ERC165 is IERC165 {
    /*
     * bytes4(keccak256('supportsInterface(bytes4)')) == 0x01ffc9a7
     */
    bytes4 private constant _INTERFACE_ID_ERC165 = 0x01ffc9a7;
```

Reply:
@dievardump nice check here. I don't think we need this level of introspection, as similar to [OZ erc721](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/ERC721.sol)(which imports their `ERC165` contract, but not extra meta, `ERC165Checker` library...), we implement the function required by ERC165 to confirm erc721 and extension support 
![image](https://user-images.githubusercontent.com/41117279/134825734-ebc5b008-8c05-4a33-a945-c4867e0aa618.png). I am curious of cases where an app or other contract might call `supportsInterface`, the function contemplated by erc165, and also wants to make sure it supports this same interface on meta level?